### PR TITLE
[CARBONDATA-200] Add performance statistics logs to record the query time taken by carbon

### PR DESCRIPTION
**Pull request by foryou2030**
# Why raise this pr?
Check query performance statistics More intuitive.
# How to resolve?
The query statistics will be shown on logs as the following, we can get performance info from the table easily:
```
+--------+--------------------+---------------------+----------------------+
|  Module|      Operation Step|                                  Query Cost|
+--------+--------------------+---------------------+----------------------+
|              JDBC connection|                     |                      |
+--------+--------------------+---------------------+----------------------+
|        |           SQL parse|                     |                1.104 |
|        +--------------------+                     +----------------------+
|        |Connect time to Hive|                     |                0.054 |
|  Driver|    metastore carbon|                     |                      |
|        +--------------------+                     +----------------------+
|    Part|Block identification|                     |                0.003 |
|        +--------------------+                     +----------------------+
|        |    Schedule task to|                     |                      |
|        |            executor|                     |                      |
+--------+--------------------+---------------------+----------------------+
|        |          Load index|                     |                0.046 |
|        +--------------------+                     +----------------------+
|Executor|   Scan(include read|                     |                0.019 |
|    Part|     data from file&|                     |                      |
|        |             process|                0.13 |                      |
|        +--------------------+                     +----------------------+
|        |     Dictionary load|                     |                0.001 |
|        +--------------------+                     +----------------------+
|        |Prepare query result|                     |                0.001 |
|        |   and give to spark|                     |                      |
+--------+--------------------+---------------------+----------------------+
|      Print result at beeline|                     |                      |
+--------+--------------------+---------------------+----------------------+
|                        Total|                     |                      |
+--------+--------------------+---------------------+----------------------+
```
Most of the statistics info can be recorded into the table, But some can't be display in the table directly, such as `JDBC connection, Schedule task, Print at beeline`. Because these statistics can't be caught in carbon code, so we need to analyzing the jdbc log and calculate.

### CODE REVIEW

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.carbondata.core.carbon.querystatistics;
+
+public final class QueryStatisticsCommonConstants {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java]: change this class to interface as in interface all the property will be by default public static final, so not need to write in the code and no need to add any constructor for the same

Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java]: Change name to QueryStatisticsConstants

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java [hunk 2]
```diff
@@ -20,11 +20,14 @@
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: Please add Driver Side block loading time in statistics

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java [hunk 3]
```diff
@@ -61,14 +67,275 @@ public QueryStatisticsRecorder(String queryId) {
...
+        // TODO: get the finished query, and print Statistics
+        if (queryStatisticsMap.get(key).size() > 8) {
+          String jdbc_connection_time = "";
+          String sql_parse_time = "";
+          String load_meta_time = "";
+          String block_identification_time = "";
+          String schedule_time = "";
+          String driver_part_time = "";
+          String executor_part_time = "";
+          String load_index_time = "";
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: change load_index_time to block loading time

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java [hunk 4]
```diff
@@ -61,14 +67,275 @@ public QueryStatisticsRecorder(String queryId) {
...
+          String scan_data_time = "";
+          String dictionary_load_time = "";
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: dictionary loading time required in both the side executor and driver, please add in driver side

**Comment by kumarvishal09** [comment]
@foryou2030 Please correct me if I am wrong, we can have multiple executor  how you are logging all the executor details, how you are printing details of each executor statistics details

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatistic.java [hunk 5]
```diff
@@ -45,8 +47,16 @@
...
 
-  public QueryStatistic() {
+  /**
+   * number of count
+   */
+  private long count;
+
+  private String queryIdWthTask;
+
+  public QueryStatistic(String queryId) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatistic.java]: Why queryId is required in QueryStatistic as queryId will be common for all the query statistics object, previously it was in query statistics recorder class as recorder was common, can we move it to recorder??

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java [hunk 6]
```diff
@@ -61,14 +67,275 @@ public QueryStatisticsRecorder(String queryId) {
    */
   public synchronized void recordStatistics(QueryStatistic statistic) {
     queryStatistics.add(statistic);
+    // refresh query Statistics Map
+    String key = statistic.getQueryId();
+    if (!StringUtils.isEmpty(key)) {
+      // 240954528274124_0 and 240954528274124 is the same query id
+      key = key.substring(0, 15);
+    }
+    if (queryStatisticsMap.get(key) != null) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: how you are handling for each executor query time ?

**Comment by foryou2030** [comment]
@kumarvishal09 i think,
new singleton QueryStatisticsRecorder for each executor, add queryStatisticsMap to QueryStatisticsRecorder , and it group by queryId, then each executor can print the statistics for same queryid togeter.
how about?

**Comment by kumarvishal09** [comment]
@gvramana @ravipesala do we need this table, or direct logging the details is ok?

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java (hunk 6 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: i think 240954528274124_0  means queryid for each segment? yes?
each executor will print whole time for all segment 
if we have 3 executors, 2 segments, print like:
executor1 log:
|        +--------------------+                     +----------------------+
|        |     Dictionary load|                     |     0.002  0.001 |
|        +--------------------+                     +----------------------+

executor 2 log:
|        +--------------------+                     +----------------------+
|        |     Dictionary load|                     |     0.003  0.002 |
|        +--------------------+                     +----------------------+

executor3 log:
|        +--------------------+                     +----------------------+
|        |     Dictionary load|                     |     0.001  0.003 |
|        +--------------------+                     +----------------------+

**Comment by kumarvishal09** [comment]
previously we were printing total number of blocks processed by each task and number of output records, please add the same in the executor statistics details

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java (hunk 6 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: query id is not based on segment it is based on task

**Comment by kumarvishal09** [comment]
@foryou2030 QueryStatisticsRecorder is singleton then it is ok, queryId will be part of statistics object

File: core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java [hunk 7]
```diff
@@ -45,10 +47,28 @@
...
+  }
+
+  private Boolean flag;
+
+  private Long total = 0L;
+
+  private QueryModel queryModel;
+
+  @Override public boolean hasNext() {
+    flag = super.hasNext();
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java]: why we are overriding hasnext?? We can handle this in super class AbstractDetailQueryResultIterator

File: core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java [hunk 8]
```diff
@@ -45,10 +47,28 @@
 
   public DetailQueryResultIterator(List<BlockExecutionInfo> infos, QueryModel queryModel) {
     super(infos, queryModel);
+    this.queryModel = queryModel;
+  }
+
+  private Boolean flag;
+
+  private Long total = 0L;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java]: Why wrapper Long is required why not primitive long ?

File: core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java [hunk 9]
```diff
@@ -45,10 +47,28 @@
...
+
+  private QueryModel queryModel;
+
+  @Override public boolean hasNext() {
+    flag = super.hasNext();
+    if(!flag && total > 0) {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java]: why total >0 check is required?

**Comment by kumarvishal09** [comment]
@foryou2030 As QueryStatisticsRecorder is a singleton class after query execution we need to remove all the statistics details for the query, please handle this

**Comment by ravipesala** [comment]
I don't think it is a proper way to show tabular format in logs. Usually it is better to have either UI or any other separate tool to show the statistics.

**Comment by foryou2030** [comment]
@kumarvishal09, after discussed with jarry and u, 
I adjust the table struct, and fixed most of the comments, the new query statics shown like the follwing:
1. use singleton QueryStatisticsRecorder record driver statistics, check print in driver log
2. use the origin QueryStatisticsRecorder  record the executor statistics for each taskid, check print in executor log.

```
STATISTIC 01-09 11:50:32,311 - Time taken to load the Block(s) In Driver Side for the taskid :  Is : 117
STATISTIC 01-09 11:50:32,358 - Time taken to load the Block(s) In Driver Side for the taskid :  Is : 3
STATISTIC 01-09 11:50:32,401 - Print query statistic for query id: 832663758913772
+--------+--------------------+---------------------+----------------------+
|  Module|      Operation Step|                                  Query Cost|
+--------+--------------------+---------------------+----------------------+
|        |           SQL parse|                     |                 1.122|
|        +--------------------+                     +----------------------+
|  Driver|      Load meta data|                 1.21|                 0.052|
|        +--------------------+                     +----------------------+
|    Part|    Block allocation|                     |                 0.033|
|        +--------------------+                     +----------------------+
|        |Block identification|                     |                 0.003|
+--------+--------------------+---------------------+----------------------+

STATISTIC 01-09 11:50:33,111 - Print query statistic for each task id:
+-----------------+----------------+--------------------+----------------+---------------+-----------+-------------------+
|          task_id|load_blocks_time|load_dictionary_time|scan_blocks_time|scan_blocks_num|result_size|total_executor_time|
+-----------------+----------------+--------------------+----------------+---------------+-----------+-------------------+
|832663758913772_0|           0.047|               0.002|           0.019|              1|        950|              0.134|
+-----------------+----------------+--------------------+----------------+---------------+-----------+-------------------+

STATISTIC 01-09 11:50:34,739 - Total Time taken to prepare query result for the taskid : 832663758913772 Is : 1
STATISTIC 01-09 11:50:34,740 - Total Time taken to prepare query result for the taskid : 832663758913772 Is : 1
```

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java (hunk 1 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java]: QueryStatisticsConstants used to defined Constants, i write it refer to CarbonCommonConstants,
i think need to change it to interface

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java (hunk 2 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsRecorder.java]: handled,  but i can't get the query id, so i print it alone

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatistic.java (hunk 5 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatistic.java]: handled

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java (hunk 1 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/QueryStatisticsCommonConstants.java]: CarbonCommonConstants also we can change it to interface

File: core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java (hunk 7 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java]: in hasnext: i need to judge when  will it traverse the iterator over
in next: i need to accumulate the total time
but AbstractDetailQueryResultIterator didn't define 'next'

File: core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java (hunk 8 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java]: handled

File: core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java (hunk 7 above)
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/scan/result/iterator/DetailQueryResultIterator.java]: you can have time and record count variable in AbstractDetailQueryResultIterator and in next you can update the value....and once has next is false you can record the statistics

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala [hunk 10]
```diff
@@ -32,6 +32,7 @@ import org.apache.carbondata.core.carbon.{AbsoluteTableIdentifier, ColumnIdentif
 import org.apache.carbondata.core.carbon.metadata.datatype.DataType
```
Comment (kumarvishal09) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: do not use System.nano time use system.currentTimeInMilii

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+      LogServiceFactory.getLogService(SingleQueryStatisticsRecorder.class.getName());
+  /**
+   * serialization version
+   */
+  private static final long serialVersionUID = -1L;
+
+  /**
+   * singleton QueryStatisticsRecorder for driver
+   */
+  private HashMap<String, List<QueryStatistic>> queryStatisticsMap;
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: use Map<String,List<QueryStatistic>>

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.HashMap;
+import java.util.List;
+
+import org.apache.carbondata.common.logging.LogService;
+import org.apache.carbondata.common.logging.LogServiceFactory;
+
+/**
+ * Class will be used to record and log the query statistics
+ */
+public class SingleQueryStatisticsRecorder implements Serializable {
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: Change class name to DriverQueryStatisticsRecorder

Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: remove Serializable as it not required

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+    return null;
+  }
+
+  /**
+   * Below method will create string like "***********"
+   *
+   * @param a
+   * @param num
+   */
+  public static String printLine(String a, int num)
```
Comment (kumarvishal09) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: move it to util class..carbonUtil

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala (hunk 10 above)
Comment (foryou2030) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonDictionaryDecoder.scala]: handled

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java (hunk 11 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: handled

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java (hunk 12 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: handled

File: core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java (hunk 13 above)
Comment (foryou2030) [inline core/src/main/java/org/apache/carbondata/core/carbon/querystatistics/SingleQueryStatisticsRecorder.java]: handled

**Comment by kumarvishal09** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/85/

**Comment by kumarvishal09** [comment]
LGTM

**Pull request closed without merging**